File: 637_AverageLevelsInBinaryTree/solution.cpp
class Solution{

public:
    vector<double> averageOfLevels(TreeNode* root){
    
        vector<int> count;
        vector<double> avgs;
        average(root, 0, count, avgs); 
        for(int i = 0; i < count.size(); i++){
            avgs[i] /= count[i];
        }
        return avgs;
    }

private:
    void average(TreeNode* root, int i, vector<int>& count, vector<int>& sum){
        if(root == null) return;
        
        if(count.size() <= i){
            count.push_back({});
        }
        if(sum.size() <= i){
            sum.push_back({});
        }
        
        count[i] += 1;
        sum[i] += root->val;
        
        average(root->left, i+1, count, sum);
        average(root->right, i+1, count, sum);
    }
}

File: 637_AverageLevelsInBinaryTree/solution.java
class Solution{
    public List<Double> averageOfLevel(TreeNode node){
    
        List<Integer> count = new ArrayList<>();
        List<Double> res = new ArrayList<>();
        average(node, 0, count, res);
        for(int i = 0; i < res.size(); i++){
            res.set(i, res.get(i) / count.get(i));
        }
        return res;
    }
    
    private void average(TreeNode node, int i, List<Double> sum, List<Integer> count){
        if(node == null) return;
    
        if(i < sum.size()){
            sum.set(i, sum.get(i) + node.val);
            count.set(i, count.get(i) + 1);
        }else{
            sum.add(node.val);
            count.add(1);
        }
        
        average(node.left, i + 1, sum, count);
        average(node.right, i + 1, sum, count);
    }
}
